puffin/include/puffin/puffin.h

```diff
@@ -0,0 +1,138 @@
+#ifndef PUFFIN_PUFFIN_H
+#define PUFFIN_PUFFIN_H
+
+#include <inttypes.h>
+#include <stdbool.h>
+#include <stdlib.h>
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+    typedef struct AGENT_TYPE *AGENT;
+
+    struct Claim;
+    typedef struct Claim Claim;
+
+    typedef void *RESULT;
+
+    typedef enum
+    {
+        RESULT_OK,
+        RESULT_IO_WOULD_BLOCK,
+        RESULT_ERROR_OTHER
+    } RESULT_CODE;
+
+    typedef struct
+    {
+        /*
+         * Any opaque data needed by the callback.
+         */
+        void *context;
+
+        /*
+         * The actual callback function, called on each claim.
+         */
+        void (*const notify)(void *context, Claim *claim);
+
+        /*
+         * Perform the necessary cleanup steps to destroy the callback.
+         */
+        void (*const destroy)(void *context);
+    } CLAIMER_CB;
+
+    typedef struct AGENT_INTERFACE
+    {
+        /*
+         * Perform cleanup tasks and release memory used by the agent.
+         */
+        void (*const destroy)(AGENT agent);
+
+        /*
+         * Process input buffer, progress internal state and optionally produce output
+         * into the output buffer.
+         */
+        RESULT (*const progress)(AGENT agent);
+
+        /*
+         * In-place reset of the agent.
+         */
+        RESULT (*const reset)(AGENT agent, uint8_t new_name);
+
+        /*
+         * Produce a textual description of the current agent's state.
+         *
+         * TODO: Remove this function? It seems
```

Comment:
Still useful for debugging (e.g., `Debug for Agent`) but I agree that using claims globally would be cleaner